net/tcp/tcp_send.c

```diff
@@ -613,8 +613,13 @@ void tcp_synack(FAR struct net_driver_s *dev, FAR struct tcp_conn_s *conn,
   tcp->optdata[optlen++] = tcp_mss & 0xff;
 
 #ifdef CONFIG_NET_TCP_WINDOW_SCALE
-  if (tcp->flags == TCP_SYN ||
-      ((tcp->flags == (TCP_ACK | TCP_SYN)) && (conn->flags & TCP_WSCALE)))
+  if (tcp_get_recvwindow(dev, conn) < UINT16_MAX)
+    {
+      conn->rcv_scale = 0;
```

Comment:
I think you should disable window scale on handshake phase:
https://github.com/apache/nuttx/blob/master/net/tcp/tcp_input.c#L601-L608

Reply:
@anchao, sorry for the late reply，seems no need to modify:
1. nuttx clisent : if set windows less 65535 in sync, the server won't set WS(window scale) in sync|ack packet，from the actual packet capture with linux.
2. nuttx server : if client set WS, we will  goto tcp_synack to reset WS

Reply:
if so, why not fill the window scale(WS) filed only if receive window greater than UINT16_MAX? is it necessary to reset rcv_scale to 0?

Reply:
I set the patch first time, but didn't reset 0 in case2 (nuttx server ) and if tcp_get_recvwindow will return 65535 when window size bigger than UINT16_MAX. so I think less than UINT16_MAX will be better understood.
```
  if ((tcp->flags == TCP_SYN ||
      ((tcp->flags == (TCP_ACK | TCP_SYN)) && (conn->flags & TCP_WSCALE))) &&
      tcp_get_recvwindow(dev, conn) >= UINT16_MAX)                                // '>' is useless
```

Reply:
How about checking `conn->rcv_bufs < UINT16_MAX` instead of `tcp_get_recvwindow(dev, conn) < UINT16_MAX`? The `rcv_bufs` is the max buffer we can get for the whole lifetime.

Reply:
@anchao what's your opinion?

Reply:
> How about checking `conn->rcv_bufs < UINT16_MAX` instead of `tcp_get_recvwindow(dev, conn) < UINT16_MAX`? The `rcv_bufs` is the max buffer we can get for the whole lifetime.

if didn't define CONFIG_NET_RECV_BUFSIZE , will have no conn->rcv_bufs and  and it will be unlimited.

Reply:
If the windows scale is migrated while connected, the behavior is undefined since `tcp_get_recvwindow()` changes dynamically.

Reply:
@Meissi-jian Have you found similar logic in other TCP stack implementations? or observed that other clients in the dump data arbitrarily change the window scale?

Reply:
Thank you very much! @anchao, how about this modification？
```
#ifdef CONFIG_NET_TCP_WINDOW_SCALE
  if (tcp->flags == TCP_SYN ||
      ((tcp->flags == (TCP_ACK | TCP_SYN)) && (conn->flags & TCP_WSCALE)))
    {
      if (tcp_get_recvwindow(dev, conn) < UINT16_MAX)
        {
          conn->rcv_scale = 0;
        }
      else
        {
          tcp->optdata[optlen++] = TCP_OPT_NOOP;
          tcp->optdata[optlen++] = TCP_OPT_WS;
          tcp->optdata[optlen++] = TCP_OPT_WS_LEN;
          tcp->optdata[optlen++] = CONFIG_NET_TCP_WINDOW_SCALE_FACTOR;
        }
    }
#endif
```

Reply:
And I found that the window size of Linux is not fixed, and it is setting accord to the size of Windows. Of course, each connection is fixed.

Reply:
Potential issue:
After rcv_scale is updated, which will not be restore to original value on current connection, right?
which will only be updated during the handshake phase.

Reply:
Yes! I read RFC1072: https://www.rfc-editor.org/rfc/rfc1072.txt , the scale factor must be set in SYNC and will not changed